ihontaryk/homework_5/dicts.py

```diff
@@ -0,0 +1,112 @@
+from collections import defaultdict
+
+# DICTS
+some_dict = {"a": 123, "b": 456, "c": 789}
+some_dict_2 = {"c": 789, "a": 123, "b": 456}
+some_dict_3 = dict.fromkeys("abc")  # {'a': None, 'b': None, 'c': None}
+wierd_dict = {"a": 1, 2: "b",
+              None: "abc",
+              (1, "a"): "some",
+              True: "kek",
+              frozenset((1, 2)): "lol"}
+
+some_dict == some_dict_2  # True
+
+# is container
+"a" in some_dict  # True
+len(some_dict)  # 3
+list(some_dict_2)  # ['c', 'a', 'b']
+
+# methods
+some_dict["a"]  # 123
+some_dict.get("a")  # 123
+some_dict.get("invalid", "default_val")  # 'default_val'
+
+some = {"a": 1, "b": 2}
+some.update({"b": "b", "c": "c"})
+copy_of_some = some.copy()  # shallow copy
+copy_of_some.clear()
+item = some.pop("a")  # {'b': 'b', 'c': 'c'}
+item = some.pop("invalid", "default_val")
+
+# setdefault - return value for key, but if key not present, create it with default value, and return value
+# same as:   item = data.get("k", "val")]; data["k"] = item
+item = some.setdefault("a", "default_for_a")
+"a" in some  # True
+some["a"] == "default_for_a"  # True
+
+item = some.setdefault("a", "other_default_for_a")
+"a" in some
+some["a"] == "default_for_a"
+
+# dict comprehension
+{k: v for k, v in some_dict.items()}
+{k * 2: str(v) for k, v in some_dict.items()}
+{k: v for k, v in some_dict.items() if k <= "b"}
+
+# index and group
+t_shirts = [{'size': 'S', 'color': 'blue', 'price': 100},
+            {'size': 'M', 'color': 'black', 'price': 200},
+            {'size': 'L', 'color': 'red', 'price': 300},
+            {'size': 'L', 'color': 'blue', 'price': 500}, ]
+
+dresses = [{'size': 'S', 'color': 'white', 'price': 400},
+           {'size': 'M', 'color': 'pink', 'price': 500},
+           {'size': 'M', 'color': 'black', 'price': 700},
+           {'size': 'L', 'color': 'yellow', 'price': 600}]
+
+skirts = [{'size': 'S', 'color': 'black', 'price': 300},
+          {'size': 'M', 'color': 'blue', 'price': 600},
+          {'size': 'M', 'color': 'pink', 'price': 500},
+          {'size': 'L', 'color': 'red', 'price': 400}]
+
+clothes_shop = [{'name': 'T-shirts', 'availability': t_shirts},
+                {'name': 'Dresses', 'availability': dresses},
+                {'name': 'Skirts', 'availability': skirts}]
+
+
+def group_by_size(clothes_shop):
+    by_size = {}
+
+    for clothes in clothes_shop:
+        for item in clothes['availability']:
+            if item['size'] not in by_size:
+                by_size[item['size']] = []
+            by_size[item['size']].append(item)
```

Comment:
```
by_size.setdefault(item['size'], []).append(item)      

# if item does not exist it will be created, added to dict and provided as return value
# if it exists it will be returned as return value
```